### PR TITLE
ARROW-8080: [C++] Add ARROW_SIMD_LEVEL option

### PR DESCRIPTION
**Pull request by frankdjx**
Default set to off.

Signed-off-by: Frank Du <frank.du@intel.com>

### CODE REVIEW

**Comment by pitrou** [comment]
I'm not sure what our policy should be on such options. On the one hand, we already have an option to enable SSE4.2 (but it's always on anyway, since most x86 CPUs less than 10 years old support it). On the other hand, passing compiler options can already be done by passing `CMAKE_CXX_FLAGS` and/org `ARROW_CXXFLAGS`, so it doesn't seem very useful to be adding dedicated options for each optimization flag.

Also, AVX512 is actually comprised of a lot of subsets (Wikipedia lists them: https://en.wikipedia.org/wiki/AVX-512#Instruction_set). It seems this option would enable the "AVX-512 Foundation" instructions, which is probably the baseline.

@emkornfield @wesm Thoughts?

**Comment by wesm** [comment]
I'm having a hard time finding clear documentation about avx512-related compiler options. Is `skylake-avx512` the only one?

File: cpp/cmake_modules/SetupCxxFlags.cmake [hunk 1]
```diff
@@ -270,6 +271,10 @@ if(CXX_SUPPORTS_SSE4_2 AND ARROW_SSE42)
   set(CXX_COMMON_FLAGS "${CXX_COMMON_FLAGS} -msse4.2")
 endif()
 
+if(CXX_SUPPORTS_AVX512 AND ARROW_AVX512)
+  set(CXX_COMMON_FLAGS "${CXX_COMMON_FLAGS} -march=skylake-avx512")
```
Comment (pitrou) [inline cpp/cmake_modules/SetupCxxFlags.cmake]: This is not the same as `-mavx512f` above.

Comment (jianxind) [reply cpp/cmake_modules/SetupCxxFlags.cmake]: skylake-avx512 is a combined options which include basic AVX512 subsets for most AVX512 capable machines. The check_cxx_compiler_flag can only inspect the single define.

echo | gcc -dM -E - -march=skylake-avx512 | grep AVX512
#define __AVX512F__ 1
#define __AVX512BW__ 1
#define __AVX512VL__ 1
#define __AVX512CD__ 1
#define __AVX512DQ__ 1

**Comment by jianxind** [comment]
> I'm having a hard time finding clear documentation about avx512-related compiler options. Is `skylake-avx512` the only one?

Yeah, skylake-avx512 is a collection option which include AVX512F, AVX512BW, AVX512VL, AVX512CD, AVX512DQ subset. These five are the most supported for latest devices(https://en.wikipedia.org/wiki/AVX-512#CPUs_with_AVX-512)

**Comment by emkornfield** [comment]
I don't feel strongly but it feels like we should be trying to keep the CMake file optimizations for the most common platform options that we need for packaging/CI, additional CPU options like this seem like they should mostly be passed through CXX_* args by expert users who need the extra performance, until we are ready to run regular tests in CI.  Even then I could see the argument of still passing them via CXX_* flags.  I'm happy to listen to reasonable arguments for why these should be included and other rubrics for deciding what to include.

**Comment by jianxind** [comment]
> I don't feel strongly but it feels like we should be trying to keep the CMake file optimizations for the most common platform options that we need for packaging/CI, additional CPU options like this seem like they should mostly be passed through CXX_* args by expert users who need the extra performance, until we are ready to run regular tests in CI. Even then I could see the argument of still passing them via CXX_* flags. I'm happy to listen to reasonable arguments for why these should be included and other rubrics for deciding what to include.

Yeah, it can pass additional options by CXX flags for expert anyway. For other new guys who don't want to play with the detail compiler AVX512 options, it could just simple the out-of-box usage if arrow has a built-in option already.

**Comment by pitrou** [comment]
I'll note that it can be a bad idea to enable AVX512, since it can cause frequency drops that aren't offset by the per-cycle performance gain:
https://reviews.llvm.org/D67259

So we should probably leave it to experts anyway and avoid lure people into enabling it.

**Comment by jianxind** [comment]
Attached the data on a "Intel® Xeon® Platinum 8280L" server with the mirco-benchmarks build by "DARROW_BUILD_BENCHMARKS", filter data shows ~70 positive(>1.1) improvements while other ~20 indicators shows regression(<0.9). Basically the overall results give a positive achievements with only leverage the compiler feature. I guess there could be more space to improve the performance on AVX512 if some critical path can be modified to use avx intrinsics API, https://software.intel.com/sites/landingpage/IntrinsicsGuide/

**Comment by pitrou** [comment]
Thank you for the benchmarks. I agree those look generally positive (with a few above 2x, which is interesting). I'm still not sure whether it's worth exposing a dedicated option. I'll let others chime in.

**Comment by pitrou** [comment]
If you have time, it would also be interesting to measure AVX2 vs AVX512 (right now your numbers compare SSE42 vs AVX512, right?).

**Comment by wesm** [comment]
We could try to make this a little more general by having an `ARROW_SIMD_LEVEL` build option (accepting nosse, sse42, avx, avx2, avx512, or something like that). Then we can set the appropriate compiler flag based on the compiler

**Comment by jianxind** [comment]
Also attached the micro benchmark data with SSE4_2, AVX2, AVX512 build option. The performance of AVX512 is best per the data while AVX2 second.

[SIMD level on CLK.xlsx](https://github.com/apache/arrow/files/4332385/SIMD.level.on.CLK.xlsx)

**Comment by pitrou** [comment]
Thank you for posting those numbers. It seems that there are very few cases where AVX512 gives significantly better results than AVX2.

**Comment by jianxind** [comment]
> We could try to make this a little more general by having an `ARROW_SIMD_LEVEL` build option (accepting nosse, sse42, avx, avx2, avx512, or something like that). Then we can set the appropriate compiler flag based on the compiler

Changed to ARROW_SIMD_LEVEL and it support NONE, SSE4_2(default), AVX2, AVX512 definition, could you help to review it again?

**Review by wesm — APPROVED** [review]
+1

**Pull request closed without merging**